app/lib/template-updater/util.js

```diff
@@ -0,0 +1,258 @@
+
+    /** @type {Template[]} */
+    let templates = [];
+
+    if (hasTemplates) {
+      try {
+        templates = JSON.parse(fs.readFileSync(templatesPath));
+
+        if (!Array.isArray(templates)) {
+          templates = [];
+        }
+      } catch (error) {
+        log.warn('Failed to parse templates', error);
+      }
+    }
+
+    const {
+      updatedTemplates,
+      warnings
+    } = await fetchAndUpdateTemplates(templates, executionPlatformVersion, url);
+
+    fs.mkdirSync(path.dirname(templatesPath), { recursive: true });
+
+    fs.writeFileSync(templatesPath, JSON.stringify(updatedTemplates, null, 2));
+
+    return { warnings, hasNew: updatedTemplates.length !== templates.length };
+  } catch (error) {
+    log.error(`Failed to update templates from ${ url }`, error);
+
+    return {
+      warnings: [ `Failed to update templates from ${ url }: ${ error.message }` ],
+      hasNew: false
+    };
+  }
+}
+
+module.exports.updateTemplates = updateTemplates;
+
+/**
+ * Fetch templates compatible with the execution platform version from URL. Skip
+ * templates that cannot be fetched or parsed without errors.
+ *
+ * @param {Template[]} connectorTemplates
+ * @param {string} executionPlatformVersion
+ * @param {string} url
+ *
+ * @returns {Promise<{ updatedTemplates: Template[], warnings: string[] }>}
+ */
+async function fetchAndUpdateTemplates(connectorTemplates, executionPlatformVersion, url) {
+  log.info('Fetching and updating templates');
+
+  let updatedTemplates = [ ...connectorTemplates ],
+      warnings = [];
+
+  let response = await fetch(url);
+
+  if (!response.ok) {
+    log.warn(`Failed to fetch templates from ${ url } (HTTP ${ response.status })`);
+
+    warnings.push(`Failed to fetch templates from ${ url } (HTTP ${ response.status })`);
+
+    return { updatedTemplates, warnings };
+  }
+
+  /** @type {TemplatesByIdMetadata} */
+  const templatesByIdMetadata = await response.json();
+
+  const stats = {
+    fetched: 0,
+    cached: 0,
+    incompatible: 0,
+    error: 0
+  };
+
+  for (const id in templatesByIdMetadata) {
+    const templatesMetadata = templatesByIdMetadata[ id ];
+
+    for (const templateMetadata of templatesMetadata) {
+      const template = updatedTemplates.find(t => t.id === id && t.version === templateMetadata.version);
+
+      if (template && isCachedRef(template, templateMetadata)) {
+        log.info(`Skipping template ${ id } version ${ templateMetadata.version } as it was already fetched from ${ templateMetadata.ref }`);
+
+        stats.cached++;
+
+        continue;
+      }
+
+      if (!isTemplateCompatible(templateMetadata, executionPlatformVersion)) {
+        log.info(`Skipping template ${ id } version ${ templateMetadata.version } as it is not compatible with execution platform version ${ executionPlatformVersion }`);
+
+        stats.incompatible++;
+
+        continue;
+      }
+
+      const { ref } = templateMetadata;
```

Comment:
Do we handle current setups of connector templates, where the `ref` is not provided?